### ISSUE TITLE
cinder_iphone_simulator scheme builds arm7/arm64 NOT i386/x86_64 via fullbuild.sh

### ISSUE THREAD

**Issue by pizthewiz** [cinder/Cinder#825]
glNext seems to build the simulator target binaries for `armv7` and `arm64`. I haven't done any iOS work with Cinder but it should be `i386` and `x86_64` no?
```sh
$  lipo -info lib/libcinder-iphone* 
Architectures in the fat file: lib/libcinder-iphone-sim.a are: armv7 arm64 
Architectures in the fat file: lib/libcinder-iphone-sim_d.a are: armv7 arm64 
Architectures in the fat file: lib/libcinder-iphone.a are: armv7 arm64 
Architectures in the fat file: lib/libcinder-iphone_d.a are: armv7 arm64 
```

I'm not familiar with the use case or specifics, but is a separate simulator target needed, would it not be enough for the iOS target to build fat?

**Comment by pizthewiz** [comment]
@richardeakin  mentioned over in #821

> @pizthewiz if you build cinder-iphone-sim with your phone still selected as the test target (even though you can't test building a static lib on a device), it will infer you need binaries for that device, instead of building for sim (x86 32bit)

The strange thing though is that this behavior might be the default as I freshly cloned the tree and just ran:
```sh
$ ./xcode/fullbuild.sh
```

**Comment by richardeakin** [comment]
I bet $10 that if you open up the xcode project and select the cinder_iphone_sim target then your phone will showing as the 'test device'. Maybe we want to fiddle with the cinder_iphone_sim target settings for 'Achitectures' and / or 'Valid Architectures'

**Comment by pizthewiz** [comment]
Strangely if I open the Xcode project to check things out, a simulator is selected and Xcode 5.1.1 on OS X 10.8.5 doesn't even seem to have an _iOS Device_ option like Xcode 6 does.

<img src="http://i.imgur.com/i3ITv2C.png">

**Comment by andrewfb** [comment]
I've added 64bit support to our iOS sim boost binaries. I'd be curious if this is still an issue in light of that...

**Comment by pizthewiz** [comment]
Building manually from Xcode and selecting a device for the _cinder_iphone_ scheme and a simulator for the _cinder_iphone_simulator_ scheme yielded legit binaries:
```sh
 ~/s/C/Cinder 🍔  lipo -info lib/*.a
Architectures in the fat file: lib/libcinder-iphone-sim_d.a are: i386 x86_64 
Architectures in the fat file: lib/libcinder-iphone_d.a are: armv7 arm64 
```

But unfortunately using `fullbuild.sh` still doesn't work:
```sh
 ~/s/C/Cinder 🍔  ./xcode/fullbuild.sh; and lipo -info lib/*.a
[…]
Architectures in the fat file: lib/libcinder-iphone-sim.a are: armv7 arm64 
Architectures in the fat file: lib/libcinder-iphone-sim_d.a are: armv7 arm64 
Architectures in the fat file: lib/libcinder-iphone.a are: armv7 arm64 
Architectures in the fat file: lib/libcinder-iphone_d.a are: armv7 arm64 
input file lib/libcinder.a is not a fat file
input file lib/libcinder_d.a is not a fat file
Non-fat file: lib/libcinder.a is architecture: x86_64
Non-fat file: lib/libcinder_d.a is architecture: x86_64
```

**Comment by pizthewiz** [comment]
Why not have a single _Cinder iOS_ target that builds fat?

**Comment by andrewfb** [comment]
I haven't researched this recently but what solution would your propose for this? A manual `lipo` at the end or is there a more modern technique? Personally I'd be open to it if we can solve the technical hurdles, and now is the time as we're hoping to merge glNext into master soon.

**Comment by pizthewiz** [comment]
Nope I was wrong - an Xcode scheme does not seem to be able to simultaneously target the `iphoneos` and `iphonesimulator` SDKs with a single build invocation. Instead, one would need to use an Aggregate target or external script that builds either a single scheme varying the SDK or building separate schemes and then `lipo` the results together.

If this were done, the application developer's experience would be a bit simplified with a single set of release/debug static libs regardless of if they were testing on device or in the simulator - the catch is that it this might require a tweak to the samples, project template and existing projects.

The core issue with `fullbuild.sh` seems to be that the SDK is either being set to or defaults to `iphoneos` for the cinder_iphone_simulator.

**Comment by richardeakin** [comment]
We'd also need to make sure you can build a valid .a from the cinder.xcodeproj itself, and that a post script wasn't strictly necessary, since this is how many people work.

There appears to be an `arch` flag to xcodebuild, though I've never tried it:

```
-arch ARCH                build each target for the architecture ARCH; this will override architectures defined in the project
```

**Comment by pizthewiz** [comment]
If we had a single `cinder_iphone` scheme, it would build valid static libs with the architecture dictated by the device vs simulator selection in the scheme selector or from the command line via the `-sdk` flag.

The problem I currently have is it is possible to create binaries targeting the simulator from the non-simulator scheme and vice versa, so I'd think it would be more clear to have one that supported both and then for distribution use a script or aggregate target to create fully-fat libs.

**Comment by pizthewiz** [comment]
Here is what I've tried locally:
- Deleted the `cinder_iphone_simulator` scheme
- `lipo`ed the simulator and device `libboost_filesystem.a` and `libboost_system.a` into `lib/ios/`
- Changed the `cinder_iphone` scheme's `VALID_ARCHS`  to `arm64 armv7 i386 x86_64`
Now the arch is drive by the device vs simulator scheme selection.

For distribution, I created a script `build-universal-ios.sh` that sits next to `fullbuild.sh`:
```bash
#!/bin/bash

CURRENT_DIR=$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )
CINDER_XCODEPROJ="${CURRENT_DIR}/cinder.xcodeproj"
CINDER_LIB_DIR=$( cd "${CURRENT_DIR}/../lib" && pwd )

CONFIGURATIONS=( Debug Release )
PRODUCT_NAMES=( cinder-iphone_d cinder-iphone )

for index in {0..1}
do
  configuration=${CONFIGURATIONS[$index]}
  product_name=${PRODUCT_NAMES[$index]}

  xcodebuild -project $CINDER_XCODEPROJ -target cinder_iphone -configuration $configuration -sdk iphonesimulator ONLY_ACTIVE_ARCH=NO PRODUCT_NAME="$product_name-iphonesimulator"
  xcodebuild -project $CINDER_XCODEPROJ -target cinder_iphone -configuration $configuration -sdk iphoneos ONLY_ACTIVE_ARCH=NO PRODUCT_NAME="$product_name-iphoneos"

  lipo -create "${CINDER_LIB_DIR}/lib$product_name-iphonesimulator.a" "${CINDER_LIB_DIR}/lib$product_name-iphoneos.a" -output "${CINDER_LIB_DIR}/$product_name.a"

  rm "${CINDER_LIB_DIR}/lib$product_name-iphonesimulator.a" "${CINDER_LIB_DIR}/lib$product_name-iphoneos.a"
done
```

This obviously leaves `fullbuild.sh` in a weird spot since it isn't quite so full, but in its current form, it really isn't full anyways.

**Comment by pizthewiz** [comment]
If that sounds good, I can PR the changes.

**Comment by richardeakin** [comment]
I don't think we can have a 'both' approach, because samples / projects / templates need to either link to the universal lib or serpatate iphone + sim libs depending on arch. It will be difficult and error prone to have to decide which one to pick per project, depending on the way cinder was built.

I'm not understanding why we can't modify the xcodebuild commands so that the resulting libcinder_iphone_sim.a is built with the flag `-arch=i386 x85_64` or whatever. It would mean we'd have to ditch the `-alltargets` approach, but i don't think that's a big deal if it means we can work around the root issu (fullbuild builds wrong arch for sim).

Don't get me wrong, I wish there were a way to not have to deal with separate iphone + sim static libraries, but I'd rather do that than require a postbuild script, as that introduces the possibility of out-of-sync build errors, which I've seen oh so many times before in projects that require postbuild steps, especially when the project is used as a sub-project dependancy.

**Comment by pizthewiz** [comment]
The issue with `fullbuild.sh` is that it doesn't specify the `-sdk` when building the iOS targets - it appears as though it falls through to `iphoneos`, so the `cinder_iphone_simulator` scheme ends up building against the wrong SDK and in turn, the wrong arch.

So we could tweak `fullbuild.sh` to instead be something like:
```sh
#!/bin/sh

CURRENT_DIR=$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )
CINDER_XCODEPROJ="${CURRENT_DIR}/cinder.xcodeproj"

# OS X
xcrun xcodebuild -project ${CINDER_XCODEPROJ} -target cinder -configuration Release $@
xcrun xcodebuild -project ${CINDER_XCODEPROJ} -target cinder -configuration Debug $@

# iOS
xcrun xcodebuild -project ${CINDER_XCODEPROJ} -target cinder_iphone -configuration Release -sdk iphoneos $@
xcrun xcodebuild -project ${CINDER_XCODEPROJ} -target cinder_iphone -configuration Debug -sdk iphoneos $@
xcrun xcodebuild -project ${CINDER_XCODEPROJ} -target cinder_iphone_simulator -configuration Release -sdk iphonesimulator $@
xcrun xcodebuild -project ${CINDER_XCODEPROJ} -target cinder_iphone_simulator -configuration Debug -sdk iphonesimulator $@
```

but the two iOS schemes in Xcode are still dependent on the device/simulator selection - ignorable for now perhaps.

**Comment by pizthewiz** [comment]
PR'd the `fullbuild.sh` tweak over in #952.

**Comment by richardeakin** [comment]
Right on, merged #952.
